### PR TITLE
fix: starting persistent web socket service (AR-3139) (RC)

### PR DESCRIPTION
**Pull request by ohassine**
<table>
<td>
  <a href="https://wearezeta.atlassian.net/browse/AR-3139" title="AR-3139" target="_blank">BugAR-3139</a>  Play Store Crash: PersistentWebSocketService.generateForegroundNotification
  </td></table>
  <br />
 

----
#### PR Submission Checklist for internal contributors

- The **PR Title**
    - [x] conforms to the style of semantic commits messages¹ supported in Wire's Github Workflow²
    - [x] contains a reference JIRA issue number like `SQPIT-764`
    - [x] answers the question: _If merged, this PR will: ..._ ³

- The **PR Description**
    - [ ] is free of optional paragraphs and you have filled the relevant parts to the best of your ability
----

# What's new in this PR?

### Issues

1. App crash when starting a foreground service.
2. Persistent webSocketService started twice from 2 different places(`GlobalObserversManager` and  `NetworkSettingsViewModel`) , which can affect battery life.
3. Non used Broadcast Receiver : `StartServiceReceiver`

### Causes (Optional)

1. Could happen when `GlobalObserversManager` tries to start a foreground service and the app is in the bakground

### Solutions

- Call startForgroundService() for Android 8 and above
- Move the logic of starting the service into one place which is the activity to be sure that the app is in the foreground.
- Remove `StartServiceReceiver`

Needs releases with:

- [ ] GitHub link to other pull request

----
#### PR Post Submission Checklist for internal contributors (Optional)

- [x] Wire's Github Workflow has automatically linked the PR to a JIRA issue
----
#### PR Post Merge Checklist for internal contributors

- [ ] If any soft of configuration variable was introduced by this PR, it has been added to the relevant documents and the CI jobs have been updated.
----
##### References
1. https://sparkbox.com/foundry/semantic_commit_messages
1. https://github.com/wireapp/.github#usage
1. E.g. `feat(conversation-list): Sort conversations by most emojis in the title #SQPIT-764`.

### CODE REVIEW

**Comment by AndroidBob** [comment]
Build [1](https://10.10.124.134/job/wire-android-reloaded/job/PR-1502/1/) **failed**.

**Comment by AndroidBob** [comment]
Build [2](https://10.10.124.134/job/wire-android-reloaded/job/PR-1502/2/) **succeeded**.

The build produced the following APK's:
- [com.wire.android-vv4.0.1(5855264)(5855264)-staging-compat.apk](https://z-lohika.s3-eu-west-1.amazonaws.com/megazord/android/reloaded/staging/compat/com.wire.android-vv4.0.1(5855264)(5855264)-staging-compat.apk)

**Comment by MohamadJaara** [comment]
Moving the service to the activity means it will start only after the app open, meaning for fdroid users have to open the app at least once after device restart

**Comment by ohassine** [comment]
> Moving the service to the activity means it will start only after the app open, meaning for fdroid users have to open the app at least once after device restart

I totally missed that point. Thanks for pointing it. 
Doing it in a global observer will crash the app since a foreground service can't be started from background starting from Android 12.
But there is some exceptions, for example when we receive ACTION_BOOT_COMPLETED
https://developer.android.com/guide/components/foreground-services#background-start-restriction-exemptions

I will revert removing the broadcast receiver to solve this

**Comment by AndroidBob** [comment]
Build [5](https://10.10.124.134/job/wire-android-reloaded/job/PR-1502/5/) **succeeded**.

The build produced the following APK's:
- [com.wire.android-vv4.0.2(5861332)(5861332)-staging-compat.apk](https://z-lohika.s3-eu-west-1.amazonaws.com/megazord/android/reloaded/staging/compat/com.wire.android-vv4.0.2(5861332)(5861332)-staging-compat.apk)

File: app/src/main/kotlin/com/wire/android/ui/WireActivityViewModel.kt [hunk 1]
```diff
@@ -403,6 +414,32 @@ class WireActivityViewModel @Inject constructor(
...
+
+                            if (statuses.any { it.isPersistentWebSocketEnabled }) {
+                                if (!servicesManager.isPersistentWebSocketServiceRunning()) {
+                                    servicesManager.startPersistentWebSocketService()
+                                }
+                            } else {
+                                servicesManager.stopPersistentWebSocketService()
+                            }
+                        }
+                    }
```
Comment (alexandreferris) [inline app/src/main/kotlin/com/wire/android/ui/WireActivityViewModel.kt:436]: this specific code related to opening incoming call screen was removed from the previous code, why was that? 🤔 

```
notificationManager.observeNotificationsAndCallsWhileRunning(
    usersToObserve,
    scope
) { call -> openIncomingCall(call.conversationId) }
```

I tried to find it somewhere but it just looks like it was deleted and not added elsewhere

**Review by alexandreferris — COMMENTED** [review]
Great work! Just have a small question

File: app/src/main/kotlin/com/wire/android/ui/WireActivityViewModel.kt [hunk 2]
```diff
@@ -403,6 +414,32 @@ class WireActivityViewModel @Inject constructor(
...
+
+                    is ObservePersistentWebSocketConnectionStatusUseCase.Result.Success -> {
+                        result.persistentWebSocketStatusListFlow.collect { statuses ->
+
+                            if (statuses.any { it.isPersistentWebSocketEnabled }) {
+                                if (!servicesManager.isPersistentWebSocketServiceRunning()) {
+                                    servicesManager.startPersistentWebSocketService()
+                                }
+                            } else {
+                                servicesManager.stopPersistentWebSocketService()
```
Comment (MohamadJaara) [inline app/src/main/kotlin/com/wire/android/ui/WireActivityViewModel.kt:433]: this stop can cause an issue if a user start and stop the web socket the app will be stuck on connecting (if the app is in the foreground)

**Review by MohamadJaara — APPROVED** [review]
LGTM, only one comment about stating and stopping will make the app stuck on connecting, but not a blocker since it is outside the scope of this fix

great work as always 👍🏼

File: app/src/main/kotlin/com/wire/android/ui/WireActivityViewModel.kt (hunk 1 above)
Comment (ohassine) [reply app/src/main/kotlin/com/wire/android/ui/WireActivityViewModel.kt:436]: it's in `PersistentWebSocketService`

File: app/src/main/kotlin/com/wire/android/ui/WireActivityViewModel.kt (hunk 2 above)
Comment (ohassine) [reply app/src/main/kotlin/com/wire/android/ui/WireActivityViewModel.kt:433]: let's tackle it in a different fix 👍🏼

**Pull request merged**